### PR TITLE
separate lockfiles for example repos

### LINKED ISSUE #94: Adding lockfiles to examples
- `examples/*` doesn't have lockfiles right now as they're part of the monorepo
- this might make the vercel deployment randomly fail if someone publishes a bad build of the 10^100 sub-deps of next (as shown yesterday in https://github.com/trpc/trpc/pull/91#issuecomment-781434382)
- would be nice if they had lockfiles but not sure how to make it happen in the yarn workspace

### PR DESCRIPTION
**Pull request by juliusmarminge**
Closes #94

## 🎯 Changes

What changes are made in this PR? Is it a feature or a bug fix?

- Runs installation with `shared-workspace-lockfile=false`
- Commits this change (this commit is not visible on the parent repo (this one), only child)
- Sync downstream using subtree

Tested here: 
- https://github.com/juliusmarminge/git-subtree-test
- https://github.com/juliusmarminge/git-subtree-test-app1
- https://github.com/juliusmarminge/git-subtree-test-app2
- https://github.com/juliusmarminge/git-subtree-test-app3

I guess the downside is that there will be a duplicate commit for the lockfile sync: 
CleanShot 2023-02-03 at 17 13 02

## ✅ Checklist

- [ ] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made.
- [ ] I have added or updated the tests related to the changes made.

### CODE REVIEW

File: .github/workflows/subtree.yml [hunk 1]
```diff
@@ -29,7 +29,42 @@ jobs:
...
+          echo "shared-workspace-lockfile=false" >> .npmrc
+
+      - name: Install dependencies
+        run: |
+          pnpm install
+
+      - name: Commit changes
+        run: |
+          git config --global user.email "email-here"
+          git config --global user.name "trpc-bot"
```
Comment (juliusmarminge) [inline .github/workflows/subtree.yml:63]: not sure what user config you have here @KATT

Comment (KATT) [reply .github/workflows/subtree.yml:48]: Does the config here matter?

Comment (juliusmarminge) [reply .github/workflows/subtree.yml]: I guess not, maybe if you want it to look like a proper commit 🤔

image

**Review by KATT — DISMISSED** [review]
The trpc deps become relative so itdoesn't fully work, unfortunately, leading the CI to fail. 

- https://github.com/trpc/examples-next-prisma-starter/blob/ec7cdebd25f1a5c6f41e55a8de90fd5cfd17158d/pnpm-lock.yaml#L39-L42
- https://github.com/trpc/examples-next-prisma-starter/tree/subtree-lockfiles
- https://github.com/trpc/examples-next-prisma-starter/actions/runs/4088954745/jobs/7051147754

**Comment by juliusmarminge** [comment]
> 
> 
> The trpc deps become relative so itdoesn't fully work, unfortunately, leading the CI to fail. 
> 
> 
> 
> - https://github.com/trpc/examples-next-prisma-starter/blob/ec7cdebd25f1a5c6f41e55a8de90fd5cfd17158d/pnpm-lock.yaml#L39-L42
> 
> - https://github.com/trpc/examples-next-prisma-starter/tree/subtree-lockfiles
> 
> - https://github.com/trpc/examples-next-prisma-starter/actions/runs/4088954745/jobs/7051147754
> 
> 

Uurgh... theres prob an option we can add here similar to the lockfile setting, https://pnpm.io/npmrc#link-workspace-packages

File: .github/workflows/subtree.yml [hunk 2]
```diff
@@ -29,7 +30,28 @@ jobs:
         with:
           fetch-depth: 0
 
-      - uses: nxtlvlsoftware/git-subtree-action@1.1
+      - uses: pnpm/action-setup@v2.2.4
+      - uses: actions/setup-node@v3
+        with:
+          node-version: 18.x
```
Comment (juliusmarminge) [inline .github/workflows/subtree.yml:36]:
```suggestion
          node-version: 18.x
          cache: 'pnpm'
```

Idk if cache Will help when it should generate a lockfile?

File: .github/workflows/subtree.yml [hunk 3]
```diff
@@ -29,7 +30,26 @@ jobs:
...
+
+      - run: echo "link-workspace-packages=false" >> .npmrc
+
+      - name: Install dependencies (create separate lockfiles)
+        run: |
+          cp pnpm-lock.yaml ./examples/${{ matrix.path }}/pnpm-lock.yaml
```
Comment (juliusmarminge) [inline .github/workflows/subtree.yml:44]: could maybe just remove workspace and run install in the example path?

```suggestion
          rm -f pnpm-workspace.yaml pnpm-lock.yaml
```

**Comment by KATT** [comment]
Found something that "works" .... but it doesn't respect the original lockfile


See diff:

- https://github.com/trpc/examples-next-prisma-starter/blob/c3ba44bda597b17d5ae08f64fd56c9ad54c0411d/pnpm-lock.yaml#L47
- https://github.com/trpc/examples-next-prisma-starter/blob/c3ba44bda597b17d5ae08f64fd56c9ad54c0411d/pnpm-lock.yaml#L47

**Comment by juliusmarminge** [comment]
> Found something that "works" .... but it doesn't respect the original lockfile
> 
> 
> 
> 
> 
> See diff:
> 
> 
> 
> - https://github.com/trpc/examples-next-prisma-starter/blob/c3ba44bda597b17d5ae08f64fd56c9ad54c0411d/pnpm-lock.yaml#L47
> 
> - https://github.com/trpc/examples-next-prisma-starter/blob/c3ba44bda597b17d5ae08f64fd56c9ad54c0411d/pnpm-lock.yaml#L47

Aah hmm the superjson version doesnt match the one in the parent repo... https://github.com/trpc/trpc/blob/4312ef45cf6141d700e1a3420c1f8483d5717744/pnpm-lock.yaml#L633

File: .github/workflows/subtree.yml (hunk 2 above)
Comment (KATT) [reply .github/workflows/subtree.yml:36]: idk

File: .github/workflows/subtree.yml [hunk 4]
```diff
@@ -29,7 +30,27 @@ jobs:
...
+      - uses: pnpm/action-setup@v2.2.4
+      - uses: actions/setup-node@v3
+        with:
+          node-version: 18.x
+
+      - name: Install dependencies and create separate lockfile
+        run: |
+          echo "shared-workspace-lockfile=false" >> .npmrc
+          echo "link-workspace-packages=false" >> .npmrc
+          cp ./pnpm-lock.yaml ./examples/${{ matrix.path }}/pnpm-lock.yaml
```
Comment (KATT) [inline .github/workflows/subtree.yml:42]: This somehow makes sure that the original pnpm lockfile is somewhat respected

**Comment by KATT** [comment]
Made a solution that is slow and very hacky... it works... but I doubt it'll work in future versions of pnpm

File: .github/workflows/subtree.yml [hunk 5]
```diff
@@ -3,6 +3,7 @@ on:
   push:
     branches:
       - main
+      - subtree-lockfiles
```
Comment (KATT) [inline .github/workflows/subtree.yml:6]:
```suggestion
```

File: .github/workflows/subtree.yml [hunk 6]
```diff
@@ -16,10 +17,10 @@ jobs:
...
             #
             next-prisma-starter,
-            next-prisma-websockets-starter,
-            next-prisma-todomvc,
-            next-minimal-starter,
-            next-big-router,
+            # next-prisma-websockets-starter,
+            # next-prisma-todomvc,
+            # next-minimal-starter,
+            # next-big-router,
```
Comment (KATT) [inline .github/workflows/subtree.yml:23]: Notetoself to revert this

**Comment by juliusmarminge** [comment]
> Made a solution that is slow and very hacky... it works... but I doubt it'll work in future versions of pnpm

Slow doesn't really matter here, it's a job only done when pushes to main - not a CI job people have to sit and wait for every push they make and wanna see if CI passes.

What makes you think it won't work on future versions? tbh I'm a bit confused why it doesn't respect it just with the `shared-lockfile` option 🤔 Might be worth opening an issue to pnpm

File: .github/workflows/subtree.yml [hunk 7]
```diff
@@ -29,7 +30,28 @@ jobs:
...
+        with:
+          node-version: 18.x
+          cache: 'pnpm'
+
+      - name: Install dependencies and create separate lockfile
+        run: |
+          echo "shared-workspace-lockfile=false" >> .npmrc
+          echo "link-workspace-packages=false" >> .npmrc
+          cp ./pnpm-lock.yaml ./examples/${{ matrix.path }}/pnpm-lock.yaml
+          pnpm install --no-frozen-lockfile
```
Comment (juliusmarminge) [inline .github/workflows/subtree.yml:44]:
```suggestion
          pnpm install --filter "./examples/${{ matrix.path }}" --filter root --no-frozen-lockfile
```

File: .github/workflows/subtree.yml [hunk 8]
```diff
@@ -29,7 +30,28 @@ jobs:
...
+        with:
+          node-version: 18.x
+          cache: 'pnpm'
+
+      - name: Install dependencies and create separate lockfile
+        run: |
+          echo "shared-workspace-lockfile=false" >> .npmrc
+          echo "link-workspace-packages=false" >> .npmrc
+          cp ./pnpm-lock.yaml ./examples/${{ matrix.path }}/pnpm-lock.yaml
+          pnpm install --filter "./examples/${{ matrix.path }}" --filter root --no-frozen-lockfile
```
Comment (juliusmarminge) [inline .github/workflows/subtree.yml:44]:
```suggestion
pnpm install --filter "./examples/${{ matrix.path }}" --filter root --no-frozen-lockfile --ignore-scripts
```

**Review by bautistaaa — APPROVED** [review]

**Pull request merged**